rpc/context.go

```diff
@@ -146,17 +146,18 @@ func (ctx *Context) SetLocalInternalServer(internalServer roachpb.InternalServer
 	ctx.localInternalServer = internalServer
 }
 
-func (ctx *Context) removeConn(key string, conn *grpc.ClientConn) {
+func (ctx *Context) removeConn(key string, meta *connMeta) {
 	ctx.conns.Lock()
 	defer ctx.conns.Unlock()
-	ctx.removeConnLocked(key, conn)
+	ctx.removeConnLocked(key, meta)
 }
 
-func (ctx *Context) removeConnLocked(key string, conn *grpc.ClientConn) {
+func (ctx *Context) removeConnLocked(key string, meta *connMeta) {
 	if log.V(1) {
 		log.Infof(ctx.masterCtx, "closing %s", key)
 	}
-	if conn != nil {
+	meta.Do(func() {}) // Make sure initialization is not in-progress.
```

Comment:
This seems problematic as it'll be reentrant on the `Sync.Once` mutex if called during error handling from the real `Do()` invocation. You should probably move this to the shutdown worker.